### ISSUE TITLE
Add wildcard support for selecting artwork (cover) file

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#374]
Original [issue 374](https://code.google.com/p/quodlibet/issues/detail?id=374) created by forumscollective on 2009-12-28T05:52:40.000Z:

I'm not sure what algorithm is presently used to ascertain which file
represents the album cover when scanning a library, but it would be really
useful to be able to specify a filename as the coverart.  As an example, in
my library, the coverart is always folder.jpg for every album, but
quodlibet on occasion picks up other jpg files as the coverart.

Implementing this would enable users to optionally specify what file should
be considered the album cover and would also possibly speed up the library
scan process (ever so slightly).

**Comment by brunob** [comment]
Copy/paste of patch proposed in https://github.com/quodlibet/quodlibet/issues/374#issuecomment-80744036 :

```diff
diff -r aa9c4c504835 quodlibet/quodlibet/formats/_audio.py
--- a/quodlibet/quodlibet/formats/_audio.py	Tue Jul 05 10:19:50 2011 +0200
+++ b/quodlibet/quodlibet/formats/_audio.py	Tue Jul 05 10:29:27 2011 +0200
@@ -588,13 +588,15 @@
             return self.get_format_cover()
 
         base = self('~dirname')
-        images = []
+        fns = []
+        add_negative = False
 
         # Issue 374: Specify artwork filename
         if cfg_getboolean("albumart", "force_filename", False):
+            add_negative = True
             path = os.path.join(base, config.get("albumart", "filename"))
-            if os.path.isfile(path):
-                images = [(100, path)]
+            for path in glob.glob(path):
+                fns.append((None, path))
         else:
             get_ext = lambda s: os.path.splitext(s)[1].lstrip('.')
 
@@ -617,33 +619,34 @@
                         if get_ext(lsub_entry) in self.__cover_exts:
                             fns.append((entry, sub_entry))
 
-            for sub, fn in fns:
-                score = 0
-                lfn = fn.lower()
-                # check for the album label number
-                if "labelid" in self and self["labelid"].lower() in lfn:
-                    score += 20
+        images = []
+        for sub, fn in fns:
+            score = 0
+            lfn = fn.lower()
+            # check for the album label number
+            if "labelid" in self and self["labelid"].lower() in lfn:
+                score += 20
 
-                # Track-related keywords
-                keywords =  [k.lower().strip() for k in [self("artist"),
-                    self("albumartist"), self("album")] if len(k) > 1]
-                score += 2 * sum(map(lfn.__contains__, keywords))
+            # Track-related keywords
+            keywords =  [k.lower().strip() for k in [self("artist"),
+                self("albumartist"), self("album")] if len(k) > 1]
+            score += 2 * sum(map(lfn.__contains__, keywords))
 
-                # Generic keywords
-                score += 3 * sum(map(lfn.__contains__, self.__cover_positive))
+            # Generic keywords
+            score += 3 * sum(map(lfn.__contains__, self.__cover_positive))
 
-                # Negatives are word-boundary-wrapped
-                # This avoids problems like "Nickelback - front"
-                negs = sum(r.search(lfn) is not None
-                           for r in self.__cover_negative_regexes)
-                score -= 2 * negs
-                #print("[%s - %s]: Album art \"%s\" scores %d (%s neg)." % (
-                #        self("artist"), self("title"), fn, score, negs))
-                if score > 0:
-                    if sub is not None:
-                        fn = os.path.join(sub, fn)
-                    images.append((score, os.path.join(base, fn)))
-            images.sort(reverse=True)
+            # Negatives are word-boundary-wrapped
+            # This avoids problems like "Nickelback - front"
+            negs = sum(r.search(lfn) is not None
+                       for r in self.__cover_negative_regexes)
+            score -= 2 * negs
+            #print("[%s - %s]: Album art \"%s\" scores %d (%s neg)." % (
+            #        self("artist"), self("title"), fn, score, negs))
+            if add_negative or score > 0:
+                if sub is not None:
+                    fn = os.path.join(sub, fn)
+                images.append((score, os.path.join(base, fn)))
+        images.sort(reverse=True)
 
         for score, path in images:
             # could be a directory
diff -r aa9c4c504835 quodlibet/quodlibet/qltk/prefs.py
--- a/quodlibet/quodlibet/qltk/prefs.py	Tue Jul 05 10:19:50 2011 +0200
+++ b/quodlibet/quodlibet/qltk/prefs.py	Tue Jul 05 10:29:27 2011 +0200
@@ -372,7 +372,8 @@
 
             entry = UndoEntry()
             entry.set_tooltip_text(
-                    _("The album art image file to use when forced"))
+                    _("The album art image file to use when forced "
+                    "(supports unix wildcards)"))
             entry.set_text(config.get("albumart", "filename"))
             entry.connect('changed', self.__changed_text, 'filename')
             # Disable entry when not forcing
diff -r aa9c4c504835 quodlibet/tests/test_formats__audio.py
--- a/quodlibet/tests/test_formats__audio.py	Tue Jul 05 10:19:50 2011 +0200
+++ b/quodlibet/tests/test_formats__audio.py	Tue Jul 05 10:29:27 2011 +0200
@@ -3,7 +3,7 @@
 import os
 
 from quodlibet import config
-from quodlibet.util import HashableDict
+from quodlibet.util import HashableDict, mkdir
 from quodlibet.formats._audio import AudioFile
 from quodlibet.formats._audio import INTERN_NUM_DEFAULT
 
@@ -452,6 +452,7 @@
                       self.full_path("nothing.jpg")
                       ]
         for f in self.files: file(f, "w").close()
+        self.dirs = []
 
     def full_path(self, path):
         return os.path.join(self.dir, path)
@@ -475,9 +476,25 @@
         for f in files:
             file(f, "w").close()
             self.files.append(f)
+
             self.failUnlessEqual(os.path.abspath(quux.find_cover().name), f)
         self.test_labelid() # labelid must work with other files present
 
+    def test_glob(self):
+        quux.sanitize()
+        config.set("albumart", "force_filename", str(True))
+        config.set("albumart", "filename", "foo/*")
+        files = [
+            self.full_path(os.path.join("foo", "back.jpg")),
+            self.full_path(os.path.join("foo", "front.jpg"))
+        ]
+        self.dirs.append(self.full_path("foo"))
+        for f in files:
+            mkdir(os.path.dirname(f))
+            file(f, "w").close()
+            self.files.append(f)
+            self.failUnless(os.path.basename(f) in quux.find_cover().name)
+
     def test_intelligent(self):
         song = quux
         song["artist"] = "Q-Man"
@@ -498,5 +515,6 @@
 
     def tearDown(self):
         map(os.unlink, self.files)
+        map(os.rmdir, self.dirs)
         config.quit()
 add(Tfind_cover)

```

**Comment by zsau** [comment]
What's the status of the globbing patch?

**Comment by declension** [comment]
@zsau any of these patches will need rewriting for the current plugin-based system, though that shouldn't be too hard.

**Comment by CreamyCookie** [comment]
Ah, so this with closed with the merge of #2814.
